File: qbx_k9unit/html/tests/testkit.js
/*
    html/tests/testkit.js

    Zero-dependency assertion + test-runner for plain Node.js, mirroring
    tests/testkit.lua's own posture on the Lua side of this codebase (see
    that file's header: "why this exists instead of busted"). Same reasons
    apply here in reverse -- no npm dependency (no jest/mocha/vitest
    install step), runs under any Node the box already has, one process
    per spec file, non-zero exit on any failure.

    USAGE (see any *_spec.js in this directory for a full example):
        const t = require('./testkit');
        t.test('description of one behavior', function () {
            t.equals(1 + 1, 2);
        });
        t.test('an async behavior', async function () {
            const result = await somePromise();
            t.isTrue(result);
        });
        t.run(); // prints summary, calls process.exit(0|1)

    Every t.test() body is queued and run in registration order inside
    t.run()'s own async loop -- a failing assertion (t.equals/t.isTrue/...
    throwing) fails only that one test case, never aborts the rest of the
    file, exactly like testkit.lua's per-test pcall.
*/
'use strict';

const cases = [];
let passed = 0;
let failed = 0;
const failures = [];

/**
 * Registers one named test case. `fn` may be sync or return a Promise
 * (async function) -- t.run() awaits either uniformly. Not executed
 * immediately; queued for t.run().
 * @param {string} name
 * @param {() => (void|Promise<void>)} fn
 */
function test(name, fn) {
    cases.push({ name, fn });
}

function fail(message) {
    const err = new Error(message);
    Error.captureStackTrace && Error.captureStackTrace(err, fail);
    throw err;
}

function formatValue(v) {
    if (typeof v === 'string') return JSON.stringify(v);
    try {
        return JSON.stringify(v);
    } catch (e) {
        return String(v);
    }
}

/** @param {*} actual @param {*} expected @param {string} [message] */
function equals(actual, expected, message) {
    if (!Object.is(actual, expected)) {
        fail(`${message ? message + ': ' : ''}expected ${formatValue(expected)}, got ${formatValue(actual)}`);
    }
}

/** @param {*} actual @param {string} [message] */
function isTrue(actual, message) {
    equals(actual, true, message);
}

/** @param {*} actual @param {string} [message] */
function isFalse(actual, message) {
    equals(actual, false, message);
}

/** @param {*} actual @param {string} [message] */
function isNull(actual, message) {
    if (actual !== null) {
        fail(`${message ? message + ': ' : ''}expected null, got ${formatValue(actual)}`);
    }
}

/** @param {*} actual @param {string} [message] */
function isUndefined(actual, message) {
    if (actual !== undefined) {
        fail(`${message ? message + ': ' : ''}expected undefined, got ${formatValue(actual)}`);
    }
}

/** @param {*} actual @param {string} [message] */
function isDefined(actual, message) {
    if (actual === undefined) {
        fail(`${message ? message + ': ' : ''}expected a defined value`);
    }
}

/** @param {string} haystack @param {string} needle @param {string} [message] */
function contains(haystack, needle, message) {
    haystack = String(haystack);
    if (!haystack.includes(needle)) {
        fail(`${message ? message + ': ' : ''}expected ${formatValue(haystack)} to contain ${formatValue(needle)}`);
    }
}

/** @param {string} haystack @param {string} needle @param {string} [message] */
function notContains(haystack, needle, message) {
    haystack = String(haystack);
    if (haystack.includes(needle)) {
        fail(`${message ? message + ': ' : ''}expected ${formatValue(haystack)} NOT to contain ${formatValue(needle)}`);
    }
}

/** @param {number} actual @param {number} expected @param {number} epsilon @param {string} [message] */
function near(actual, expected, epsilon, message) {
    if (typeof actual !== 'number' || !isFinite(actual) || Math.abs(actual - expected) > epsilon) {
        fail(`${message ? message + ': ' : ''}expected ${formatValue(actual)} to be within ${epsilon} of ${formatValue(expected)}`);
    }
}

/**
 * Fails unless `fn` throws (sync) -- used sparingly; most of app.js is
 * built to never throw (its whole "graceful degradation" contract), so
 * most specs assert the ABSENCE of a throw, not its presence. Kept for the
 * few real "should throw" cases (none currently expected in app.js's own
 * contract, provided for completeness/parity with a normal assertion lib).
 * @param {() => void} fn @param {string} [message]
 */
function throws(fn, message) {
    let didThrow = false;
    try {
        fn();
    } catch (e) {
        didThrow = true;
    }
    if (!didThrow) {
        fail(`${message ? message + ': ' : ''}expected function to throw`);
    }
}

/** Runs every queued case in order, prints PASS/FAIL lines + a summary,
 * and calls process.exit() with 0 (all passed) or 1 (>=1 failed) -- same
 * contract as testkit.lua's M.summary()/os.exit() pairing, so tests/run.sh's
 * sibling in this directory can aggregate exit codes identically. */
async function run() {
    for (const { name, fn } of cases) {
        try {
            await fn();
            passed++;
            console.log(`  [PASS] ${name}`);
        } catch (err) {
            failed++;
            failures.push({ name, err });
            console.log(`  [FAIL] ${name} -- ${err && err.message ? err.message : err}`);
        }
    }

    console.log('');
    console.log(`${passed} passed, ${failed} failed`);
    if (failed > 0) {
        console.log('');
        console.log('Failures:');
        for (const f of failures) {
            console.log(`  - ${f.name}: ${f.err && f.err.message ? f.err.message : f.err}`);
        }
    }

    process.exit(failed === 0 ? 0 : 1);
}

module.exports = {
    test,
    equals,
    isTrue,
    isFalse,
    isNull,
    isUndefined,
    isDefined,
    contains,
    notContains,
    near,
    throws,
    fail,
    run,
};

File: qbx_k9unit/tests/testkit.lua
--[[
    tests/testkit.lua

    Zero-dependency assertion + test-runner for plain lua5.4. See
    DEVELOPER_REFERENCE.md for why this exists instead of busted (not installed
    for Lua 5.4 in this environment -- only for Lua 5.1 via luarocks).

    USAGE (see any *_spec.lua for a full example):
        local t = dofile('testkit.lua')
        t.test('description of one behavior', function()
            t.equals(1 + 1, 2)
        end)
        os.exit(t.summary())

    Every t.test() body runs inside its own pcall -- one failing assertion
    (t.equals/t.isTrue/... calling error()) fails only that one test, never
    aborts the rest of the file.
]]

local M = {}

M.passed = 0
M.failed = 0
M.failures = {}

--- Runs `fn` as one named test case. Never throws -- a failure inside `fn`
--- (an assertion error, or any other Lua error) is caught and recorded.
--- @param name string
--- @param fn fun()
function M.test(name, fn)
    local ok, err = xpcall(fn, debug.traceback)
    if ok then
        M.passed = M.passed + 1
        print(('  [PASS] %s'):format(name))
    else
        M.failed = M.failed + 1
        M.failures[#M.failures + 1] = { name = name, err = err }
        print(('  [FAIL] %s -- %s'):format(name, tostring(err)))
    end
end

--- @param actual any
--- @param expected any
--- @param message string?
function M.equals(actual, expected, message)
    if actual ~= expected then
        error(('%sexpected %s, got %s'):format(
            message and (message .. ': ') or '', tostring(expected), tostring(actual)
        ), 2)
    end
end

--- @param actual any
--- @param message string?
function M.isTrue(actual, message)
    M.equals(actual, true, message)
end

--- @param actual any
--- @param message string?
function M.isFalse(actual, message)
    M.equals(actual, false, message)
end

--- @param actual any
--- @param message string?
function M.isNil(actual, message)
    if actual ~= nil then
        error(('%sexpected nil, got %s'):format(
            message and (message .. ': ') or '', tostring(actual)
        ), 2)
    end
end

--- @param actual any
--- @param message string?
function M.isNotNil(actual, message)
    if actual == nil then
        error(('%sexpected a non-nil value'):format(message and (message .. ': ') or ''), 2)
    end
end

--- Substring containment check -- used throughout the admin.lua/progression.lua
--- specs to assert on captured print()/notify output without depending on
--- exact formatting.
--- @param haystack string
--- @param needle string
--- @param message string?
function M.contains(haystack, needle, message)
    haystack = tostring(haystack)
    if not haystack:find(needle, 1, true) then
        error(('%sexpected %q to contain %q'):format(
            message and (message .. ': ') or '', haystack, needle
        ), 2)
    end
end

--- @param haystack string
--- @param needle string
--- @param message string?
function M.notContains(haystack, needle, message)
    haystack = tostring(haystack)
    if haystack:find(needle, 1, true) then
        error(('%sexpected %q NOT to contain %q'):format(
            message and (message .. ': ') or '', haystack, needle
        ), 2)
    end
end

--- Prints the final pass/fail tally and returns a process exit code (0 if
--- everything passed, 1 otherwise) -- pass straight to os.exit().
--- @return integer exitCode
function M.summary()
    print('')
    print(('%d passed, %d failed'):format(M.passed, M.failed))
    if M.failed > 0 then
        print('')
        print('Failures:')
        for _, failure in ipairs(M.failures) do
            print(('  - %s: %s'):format(failure.name, tostring(failure.err)))
        end
    end
    return M.failed == 0 and 0 or 1
end

return M
